### PR TITLE
Basic DQN setup

### PR DESCRIPTION
**Pull request by ben-weiss0**
This has a lot of the work for basic DQN setup. Things that still need to be done:
- add each episode to memory
- compute Q values
- overall, better ways to build and learn
- right now car 0 is hooked into the DQN, but it is not acquiring a ton of rewards

### CODE REVIEW

File: gym_multi_car_racing/multi_car_racing.py [hunk 1]
```diff
@@ -64,6 +67,22 @@
 
 ROAD_COLOR = [0.4, 0.4, 0.4]
 
+# Replay memory parameters
+REPLAY_MEMORY_SIZE = 10000  # Maximum size of the replay memory
+BATCH_SIZE = 64             # Size of training batches sampled from memory
+replay_memory = deque(maxlen=REPLAY_MEMORY_SIZE)
```
Comment (ben-weiss0) [inline gym_multi_car_racing/multi_car_racing.py:76]: Memory of past episodes is a deque

File: gym_multi_car_racing/multi_car_racing.py [hunk 2]
```diff
@@ -64,6 +67,22 @@
...
+# Exploration parameters
+EPSILON_START = 1.0       # Initial exploration rate
+EPSILON_END = 0.1         # Final exploration rate
+EPSILON_DECAY = 0.995     # Decay rate for exploration rate
+epsilon = EPSILON_START
+
+# Hyperparameters
+GAMMA = 0.99                 # Discount factor for future rewards
+LEARNING_RATE = 1e-4         # Learning rate for optimizer
+LOW_REWARD_THRESHOLD = -10
```
Comment (ben-weiss0) [inline gym_multi_car_racing/multi_car_racing.py:87]: Current number where program restarts and bails on current episode. Can be adjusted as we go.

File: gym_multi_car_racing/multi_car_racing.py [hunk 3]
```diff
@@ -716,24 +767,44 @@ def key_release(k, mod):
...
+            steps = 0
+            restart = False
+            while True:
+                s, r, done, info = env.step(a)
+                learning_state = env.state[0]  # Use car 0's state
+                state_tensor = torch.tensor(learning_state, dtype=torch.float32).permute(2, 0, 1).unsqueeze(0)
+            
+                with torch.no_grad():
+                    # Case we use best value
+                    if random.random() > epsilon:
```
Comment (ben-weiss0) [inline gym_multi_car_racing/multi_car_racing.py]: Determine if random action should be taken or not

File: gym_multi_car_racing/multi_car_racing.py [hunk 4]
```diff
@@ -716,24 +767,44 @@ def key_release(k, mod):
...
+                    # Case we do something completely random
+                    else:
+                        action_values = [random.uniform(-1, 1), random.uniform(0, 1), random.uniform(0, 1)]
+
+                # Decay epsilon if not at determined minimum
+                if epsilon > EPSILON_END:
+                    epsilon *= EPSILON_DECAY
+
+                # Define a mapping for the DQN action outputs to [steer, gas, brake] values
+                a[0] = action_values
```
Comment (ben-weiss0) [inline gym_multi_car_racing/multi_car_racing.py]: Input being given to car 0

File: gym_multi_car_racing/multi_car_racing.py [hunk 5]
```diff
@@ -716,24 +767,44 @@ def key_release(k, mod):
...
+            
+                # Process the chosen action
+                total_reward += r
+                if total_reward[0] < LOW_REWARD_THRESHOLD:
+                    restart = True
```
Comment (ben-weiss0) [inline gym_multi_car_racing/multi_car_racing.py]: Restart if reward for car 0 drops below threshold set
